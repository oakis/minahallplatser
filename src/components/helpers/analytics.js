import analytics from '@react-native-firebase/analytics';

const convertSpaces = str => str.replace(/ /g, '_');

export const track = (title, props) => {
  if (props) {
    window.log(`Firebase analytics track with props: ${title}`, props);
    analytics().logEvent(convertSpaces(title), props);
  } else {
    window.log(`Firebase analytics track: ${title}`);
    analytics().logEvent(convertSpaces(title), {});
  }
};
