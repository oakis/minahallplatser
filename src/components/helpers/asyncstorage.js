import { AsyncStorage } from 'react-native';

export const generateUid = () => {
    return `${s4()}${s4()}-${s4()}-${s4()}-${s4()}-${s4()}${s4()}${s4()}`;
};

function s4() {
    return Math.floor((1 + Math.random()) * 0x10000).toString(16).substring(1);
}

/**
 * @param {string} item 
 * @returns {Promise} Loaded storage
 */
export const getStorage = (item) => {
    return new Promise((resolve, reject) => {
        AsyncStorage.getItem(item)
        .then((json) => {
            const data = json === null ? {} : JSON.parse(json);
            window.log(`getStorage() ${item}: OK`, data);
            resolve(data);
        })
        .catch((e) => {
            window.log(`getStorage() ${item}: FAIL`, e);
            reject(e);
        });
    });
};

/**
 * @param {string} item 
 * @param {JSON} json 
 * @returns {Promise} -
 */
export const setStorage = (item, json) => {
    return new Promise((resolve, reject) => {
        AsyncStorage.setItem(item, JSON.stringify(json))
        .then(() => {
            window.log(`setStorage() ${item}: OK`);
            resolve();
        })
        .catch((e) => {
            window.log(`setStorage() ${item}: FAIL`, e);
            reject(e);
        });
    });
};
