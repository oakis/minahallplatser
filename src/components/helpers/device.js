import { Platform } from 'react-native';

export const isAndroid = () => {
    return Platform.OS === 'android';
};

export const isIOS = () => {
    return Platform.OS === 'ios';
};
