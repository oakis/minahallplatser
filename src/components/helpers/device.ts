import {Platform} from 'react-native';
import {
  getModel,
  getSystemVersion,
  getVersion,
  getUniqueId,
} from 'react-native-device-info';

export const isAndroid = (): boolean => Platform.OS === 'android';

export const isIOS = (): boolean => Platform.OS === 'ios';

export const getDeviceModel = (): string => getModel();

export const getOsVersion = (): string => getSystemVersion();

export const getAppVersion = (): string => getVersion();

export const deviceId = (): string => getUniqueId();
