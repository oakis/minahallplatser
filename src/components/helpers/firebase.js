import fetch from 'react-native-cancelable-fetch';
import { firebaseFunctionsUrl } from '../../Server';
import { handleJsonFetch } from './';

export const updateDeparturesCount = (count) => {
    if (__DEV__) {
        return;
    }
    const url = `${firebaseFunctionsUrl}/addDeparturesCount?count=${count}`;
    fetch(url, {}, 'updateDeparturesCount')
    .finally(handleJsonFetch)
    .then(({ message }) => {
        window.log(`updateDeparturesCount(): OK (${message})`);
    })
    .catch((err) => {
        window.log(`updateDeparturesCount(): FAILED' ${message})`);
        new Error(err);
    });
};

export const getDeparturesCount = () => {
    const url = `${firebaseFunctionsUrl}/getDeparturesCount`;
    return fetch(url, {}, 'getDeparturesCount')
    .finally(handleJsonFetch)
    .then(({ departuresCount }) => {
        window.log('getDeparturesCount(): OK', departuresCount);
        return departuresCount;
    })
    .catch((err) => {
        window.log('getDeparturesCount(): FAILED', err);
        new Error(err);
    });
};

export const updateStopsCount = () => {
    if (__DEV__) {
        return;
    }
    const url = `${firebaseFunctionsUrl}/addStopsCount`;
    fetch(url, {}, 'updateStopsCount')
    .finally(handleJsonFetch)
    .then(({ message }) => {
        window.log('updateStopsCount(): OK', message);
    })
    .catch((err) => {
        window.log('updateStopsCount(): FAILED', message);
        new Error(err);
    });
};

export const getStopsCount = () => {
    const url = `${firebaseFunctionsUrl}/getStopsCount`;
    return fetch(url, {}, 'getStopsCount')
    .finally(handleJsonFetch)
    .then(({ stopsCount }) => {
        window.log('getStopsCount(): OK', stopsCount);
        return stopsCount;
    })
    .catch((err) => {
        window.log('getStopsCount(): FAILED', err);
        new Error(err);
    });
};

export const incrementStopsOpened = (user, stopId) => {
    const url = `${firebaseFunctionsUrl}/incrementStopsOpen?user=${user}&stopId=${stopId}`;
    fetch(url, {}, 'incrementStopsOpened')
    .finally(handleJsonFetch)
    .then(({ message }) => {
        window.log('incrementStopsOpened(): OK', message);
    })
    .catch((err) => {
        window.log('incrementStopsOpened(): FAILED', err);
        new Error(err);
    });
}