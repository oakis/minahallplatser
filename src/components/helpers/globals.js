export const globals = {
    shouldExitApp: false,
    didLogout: false,
    isLoggingIn: false
};
