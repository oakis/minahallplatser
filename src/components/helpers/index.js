export * from './device';
export * from './message';
export * from './network';
export * from './token';
export * from './firebase';
export * from './analytics';
export * from './globals';
export * from './image';

export const isObject = (obj) => {
    return Object.prototype.toString.call(obj) === '[object Object]';
};
