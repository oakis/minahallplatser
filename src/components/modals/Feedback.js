import React, {useState} from 'react';
import {View, Modal, ScrollView, Alert} from 'react-native';
import {Button, ListHeading, Input, Text} from '@common';
import {metrics, colors} from '@style';
import {
  track,
  getDeviceModel,
  getOsVersion,
  getAppVersion,
  handleJsonFetch,
} from '../helpers';

const inputStyle = {
  borderRadius: 15,
  paddingLeft: metrics.margin.sm,
  paddingRight: metrics.margin.sm,
  marginTop: metrics.margin.md,
  marginBottom: metrics.margin.md,
  backgroundColor: '#fff',
};

export const Feedback = props => {
  const [loading, setLoading] = useState(false);
  const [name, setName] = useState('');
  const [email, setEmail] = useState('');
  const [message, setMessage] = useState('');
  const [validated, setValidated] = useState(true);

  const onChangeEmail = str => {
    setEmail(str);
    setValidated(true);
  };

  const onChangeName = str => {
    setName(str);
    setValidated(true);
  };

  const onChangeMessage = str => {
    setMessage(str);
    setValidated(true);
  };

  const cancel = () => {
    track('Feedback Cancel');
    props.close();
  };

  const reset = () => {
    setLoading(false);
    setName('');
    setEmail('');
    setMessage('');
    setValidated(true);
  };

  const sendFeedback = () => {
    if (validate()) {
      setLoading(true);
      window.log(
        `Send feedback: - Name: ${name} - E-mail: ${email} - Message: ${message} - Device: ${getDeviceModel()} - OS: ${getOsVersion()} - App Version: ${getAppVersion()}`,
      );
      const url = `${
        process.env.FIREBASE
      }/sendFeedback?name=${name}&email=${email}&message=${message}&device=${getDeviceModel()}&os=${getOsVersion()}&appVersion=${getAppVersion()}`;
      fetch(url, {}, 'sendFeedback')
        .then(handleJsonFetch)
        .then(() => {
          track('Feedback Send');
          window.log('sendFeedback(): OK');
          Alert.alert('', 'Tack för din feedback!');
          reset();
        })
        .catch(err => {
          track('Feedback Failed', {Error: err});
          window.log('sendFeedback(): FAILED', err);
          Alert.alert('', 'Något gick snett, försök igen senare.');
          setLoading(false);
        })
        .then(props.close);
    } else {
      setValidated(false);
    }
  };

  const validate = () => {
    return name.length > 0 && message.length > 0 && email.length > 0;
  };

  return (
    <Modal
      visible={props.visible}
      onRequestClose={props.close}
      animationType="slide">
      <View style={{flex: 1, backgroundColor: colors.background}}>
        <ScrollView
          contentContainerStyle={{flex: 0}}
          keyboardShouldPersistTaps="always">
          <ListHeading
            text="Skicka feedback"
            style={{marginTop: metrics.margin.lg}}
          />
          <View style={{padding: metrics.padding.md, width: '100%'}}>
            <Text>
              Namn <Text style={{color: colors.danger}}>*</Text>
            </Text>
            <Input
              value={name}
              onChangeText={onChangeName}
              style={inputStyle}
              underlineColorAndroid="#fff"
            />
            <Text>
              E-mail <Text style={{color: colors.danger}}>*</Text>
            </Text>
            <Input
              value={email}
              onChangeText={onChangeEmail}
              style={inputStyle}
              underlineColorAndroid="#fff"
            />
            <Text>
              Meddelande <Text style={{color: colors.danger}}>*</Text>
            </Text>
            <Input
              value={message}
              onChangeText={onChangeMessage}
              style={inputStyle}
              underlineColorAndroid="#fff"
              multiline
            />
            {validated ? null : (
              <Text
                style={{
                  color: colors.danger,
                  marginBottom: metrics.margin.md,
                }}>
                Var god fyll i ditt namn, en giltig e-mail och ett meddelande.
              </Text>
            )}
            <Button
              label="Skicka feedback"
              onPress={sendFeedback}
              uppercase
              loading={loading}
              color="primary"
              fontColor="alternative"
            />
            <Button
              label="Avbryt"
              onPress={cancel}
              uppercase
              color="danger"
              fontColor="alternative"
            />
          </View>
        </ScrollView>
      </View>
    </Modal>
  );
};
