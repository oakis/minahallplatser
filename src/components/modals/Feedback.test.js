import React from 'react';
import { shallow } from 'enzyme';
import firebase from 'react-native-firebase';
import { Feedback } from './Feedback';
import { track } from '../helpers';

it('should match snapshot', () => {
    const wrapper = shallow(<Feedback />);
    expect(wrapper).toMatchSnapshot();
});

it('reset() should reset to initial state', () => {
    firebase.auth = () => ({ currentUser: { email: 'abc@123.com' } });
    const wrapper = shallow(<Feedback />);
    const initialState = wrapper.state();
    wrapper.setState({
        loading: true,
        name: 'abc',
        email: 'qwerty@asd.com',
        message: 'asdasd',
        validated: false
    });
    wrapper.instance().reset();
    expect(wrapper.state()).toEqual(initialState);
});

it('validate() should return true if all fields are filled in', () => {
    const wrapper = shallow(<Feedback />);
    expect(wrapper.instance().validate()).toBe(false);
    wrapper.setState({
        name: 'abc',
        email: 'abc@123.com',
        message: 'asdasd',
    });
    expect(wrapper.instance().validate()).toBe(true);
});

it('typing name should fire onChangeName and set correct state', () => {
    const wrapper = shallow(<Feedback />);
    const input = wrapper.find('Input').at(0);
    input.simulate('changeText', 'My Name');
    expect(wrapper.state()).toEqual(expect.objectContaining({ name: 'My Name', validated: true }));
});

it('typing email should fire onChangeEmail and set correct state', () => {
    const wrapper = shallow(<Feedback />);
    const input = wrapper.find('Input').at(1);
    input.simulate('changeText', 'my@email.com');
    expect(wrapper.state()).toEqual(expect.objectContaining({ email: 'my@email.com', validated: true }));
});

it('typing message should fire onChangeMessage and set correct state', () => {
    const wrapper = shallow(<Feedback />);
    const input = wrapper.find('Input').at(2);
    input.simulate('changeText', 'my message');
    expect(wrapper.state()).toEqual(expect.objectContaining({ message: 'my message', validated: true }));
});

describe('press avbryt button', () => {
    let wrapper;
    const close = jest.fn();
    beforeEach(() => {
        track.mockClear();
        close.mockClear();
        wrapper = shallow(<Feedback close={close} />);
        wrapper.find('Button').at(1).simulate('press');
    });

    it('should fire this.props.close', () => {
        expect(close).toHaveBeenCalledTimes(1);
    });

    it('should be tracked', () => {
        expect(track).toBeCalledWith('Feedback Cancel');
    });
});

describe('press skicka feedback button', () => {
    let wrapper;
    beforeEach(() => {
        track.mockClear();
        wrapper = shallow(<Feedback />).setState({ name: 'A', email: 'B', message: 'C' });
        wrapper.instance().setState = jest.fn();
        wrapper.find('Button').at(0).simulate('press');
    });

    it('should indicate loading', () => {
        expect(wrapper.instance().setState).toBeCalledWith({ loading: true });
    });
});
