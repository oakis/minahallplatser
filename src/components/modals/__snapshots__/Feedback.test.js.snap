// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  visible={true}
>
  <View
    style={
      Object {
        "backgroundColor": "#EAE9EF",
        "flex": 1,
      }
    }
  >
    <RCTScrollView
      contentContainerStyle={
        Object {
          "flex": 0,
        }
      }
      keyboardShouldPersistTaps="always"
    >
      <View>
        <View>
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "height": 29,
                  "justifyContent": "center",
                },
                Object {
                  "marginTop": 20,
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "fontFamily": "System",
                    "fontSize": 16,
                  },
                  Object {
                    "fontSize": 20,
                    "fontWeight": "bold",
                    "marginTop": 10,
                    "paddingLeft": 10,
                    "paddingRight": 5,
                  },
                  Object {
                    "height": 29,
                  },
                ]
              }
            >
              Skicka feedback
            </Text>
            <View
              style={
                Object {
                  "flex": 1,
                  "height": 29,
                  "justifyContent": "center",
                  "marginTop": 10,
                  "paddingRight": 10,
                }
              }
            >
              <MaterialIcon
                size={24}
                style={
                  Object {
                    "textAlign": "right",
                  }
                }
              />
            </View>
          </View>
          <View
            style={
              Object {
                "backgroundColor": "#779ECB",
                "height": 2,
                "width": "100%",
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "padding": 10,
              "width": "100%",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "System",
                  "fontSize": 16,
                },
                undefined,
              ]
            }
          >
            Namn 
            <Text
              style={
                Array [
                  Object {
                    "fontFamily": "System",
                    "fontSize": 16,
                  },
                  Object {
                    "color": "#FF6961",
                  },
                ]
              }
            >
              *
            </Text>
          </Text>
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "marginBottom": 10,
                  "shadowColor": "#282828",
                  "shadowOffset": Object {
                    "height": 0,
                    "width": 0,
                  },
                  "shadowOpacity": 0.2,
                  "shadowRadius": 5,
                },
                Object {
                  "backgroundColor": "#fff",
                  "borderRadius": 15,
                  "marginBottom": 10,
                  "marginTop": 10,
                  "paddingLeft": 5,
                  "paddingRight": 5,
                },
              ]
            }
          >
            <TextInput
              autoCapitalize="none"
              autoCorrect={false}
              autoFocus={false}
              multiline={false}
              onChangeText={[Function]}
              secureTextEntry={false}
              style={
                Object {
                  "flex": 1,
                  "marginTop": 0,
                }
              }
              underlineColorAndroid="#fff"
              value=""
            />
          </View>
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "System",
                  "fontSize": 16,
                },
                undefined,
              ]
            }
          >
            E-mail 
            <Text
              style={
                Array [
                  Object {
                    "fontFamily": "System",
                    "fontSize": 16,
                  },
                  Object {
                    "color": "#FF6961",
                  },
                ]
              }
            >
              *
            </Text>
          </Text>
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "marginBottom": 10,
                  "shadowColor": "#282828",
                  "shadowOffset": Object {
                    "height": 0,
                    "width": 0,
                  },
                  "shadowOpacity": 0.2,
                  "shadowRadius": 5,
                },
                Object {
                  "backgroundColor": "#fff",
                  "borderRadius": 15,
                  "marginBottom": 10,
                  "marginTop": 10,
                  "paddingLeft": 5,
                  "paddingRight": 5,
                },
              ]
            }
          >
            <TextInput
              autoCapitalize="none"
              autoCorrect={false}
              autoFocus={false}
              multiline={false}
              onChangeText={[Function]}
              secureTextEntry={false}
              style={
                Object {
                  "flex": 1,
                  "marginTop": 0,
                }
              }
              underlineColorAndroid="#fff"
              value=""
            />
          </View>
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "System",
                  "fontSize": 16,
                },
                undefined,
              ]
            }
          >
            Meddelande 
            <Text
              style={
                Array [
                  Object {
                    "fontFamily": "System",
                    "fontSize": 16,
                  },
                  Object {
                    "color": "#FF6961",
                  },
                ]
              }
            >
              *
            </Text>
          </Text>
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "marginBottom": 10,
                  "shadowColor": "#282828",
                  "shadowOffset": Object {
                    "height": 0,
                    "width": 0,
                  },
                  "shadowOpacity": 0.2,
                  "shadowRadius": 5,
                },
                Object {
                  "backgroundColor": "#fff",
                  "borderRadius": 15,
                  "marginBottom": 10,
                  "marginTop": 10,
                  "paddingLeft": 5,
                  "paddingRight": 5,
                },
              ]
            }
          >
            <TextInput
              autoCapitalize="none"
              autoCorrect={false}
              autoFocus={false}
              multiline={true}
              onChangeText={[Function]}
              secureTextEntry={false}
              style={
                Object {
                  "flex": 1,
                  "marginTop": 0,
                }
              }
              underlineColorAndroid="#fff"
              value=""
            />
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "alignSelf": "stretch",
                "backgroundColor": "#779ECB",
                "borderRadius": 3,
                "elevation": 1,
                "flexDirection": "row",
                "justifyContent": "center",
                "marginBottom": 5,
                "opacity": 1,
                "paddingBottom": 10,
                "paddingTop": 10,
                "shadowColor": "#282828",
                "shadowOffset": Object {
                  "height": 0,
                  "width": 0,
                },
                "shadowOpacity": 0.5,
                "shadowRadius": 1,
              }
            }
          >
            <MaterialIcon
              name=""
              size={24}
            />
            <Text
              style={
                Array [
                  Object {
                    "fontFamily": "System",
                    "fontSize": 16,
                  },
                  Object {
                    "color": "#fff",
                    "fontFamily": "System",
                    "fontWeight": "bold",
                    "marginLeft": 0,
                  },
                ]
              }
            >
              SKICKA FEEDBACK
            </Text>
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "alignSelf": "stretch",
                "backgroundColor": "#FF6961",
                "borderRadius": 3,
                "elevation": 1,
                "flexDirection": "row",
                "justifyContent": "center",
                "marginBottom": 5,
                "opacity": 1,
                "paddingBottom": 10,
                "paddingTop": 10,
                "shadowColor": "#282828",
                "shadowOffset": Object {
                  "height": 0,
                  "width": 0,
                },
                "shadowOpacity": 0.5,
                "shadowRadius": 1,
              }
            }
          >
            <MaterialIcon
              name=""
              size={24}
            />
            <Text
              style={
                Array [
                  Object {
                    "fontFamily": "System",
                    "fontSize": 16,
                  },
                  Object {
                    "color": "#fff",
                    "fontFamily": "System",
                    "fontWeight": "bold",
                    "marginLeft": 0,
                  },
                ]
              }
            >
              AVBRYT
            </Text>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</Modal>
`;
